### PR TITLE
Azure Core logging interceptors

### LINKED ISSUE #127: Create logger interface and implement HTTP Pipeline logging policies
We've a ClientLogger contract defined in Java azure-core, we should be able to reuse most of it. It's pluggable for SLF4J adapters. There exists an slf4j-android `org.slf4j:slf4j-android` which by default forward log calls to https://developer.android.com/reference/android/util/Log.html

  - [x] Logger interface
  - [x] HTTP logger policy
  - [x] cURL-formatted logger policy

### PR DESCRIPTION
**Pull request by vcolin7**
Thanks for taking the time to contribute. Please take a moment to fill out the information below.

Partially fixes #127.

Changes Proposed in this pull request:
- Create LoggingInterceptor and some helper artifacts
- **(PENDING)** Create interceptor for logging request in a cURL format
- Create artifacts required to parse URLs

### CODE REVIEW

File: sdk/build.gradle [hunk 1]
```diff
@@ -6,8 +6,12 @@ buildscript {
         jcenter()
     }
 
+    ext {
```
Comment (vcolin7) [inline sdk/build.gradle]: These same changes are being included in one of my other PRs (#128).

File: sdk/core/azure-core/build.gradle [hunk 2]
```diff
@@ -1,4 +1,4 @@
-apply plugin: 'com.android.library'
+apply plugin: "com.android.library"
```
Comment (vcolin7) [inline sdk/core/azure-core/build.gradle]: These same changes are being included in one of my other PRs (#128).

File: sdk/core/azure-core/src/main/java/com/azure/android/core/http/interceptor/LoggingInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+    public Response intercept(@NonNull Chain chain) throws IOException {
+        Request request = chain.request();
+        Response response;
+
+        // No logging will be performed, trigger a no-op.
+        if (logDetailLevel == LogDetailLevel.NONE) {
+            return chain.proceed(request);
+        }
+
+        final ClientLogger logger = new ClientLogger(request.method() + " " + getRedactedUrl(request.url()));
```
Comment (vcolin7) [inline sdk/core/azure-core/src/main/java/com/azure/android/core/http/interceptor/LoggingInterceptor.java]: In the Java SDK a context parameter called "caller-method" is logged here. In absence of a context object, for now I'm logging the HTTP method type and the URL. Need to talk to someone on the java team to figure what this message looks like.

**Comment by vcolin7** [comment]
At this stage, should we consider a way of checking what the system/environment logging level is? PR #140 is not checking for this and for now I have not included any logic for this as well.

**Comment by bsiegel** [comment]
> At this stage, should we consider a way of checking what the system/environment logging level is? PR #140 is not checking for this and for now I have not included any logic for this as well.

I don't believe that's a concept on Android, because similar to iOS, all logging messages are always generated (except for Debug level message in release builds). The log viewing tool (for Android, this is `logcat`) is used to filter messages at the log level desired by the user of the tool.

**Comment by vcolin7** [comment]
Fixed a couple issues related to LoggingInterceptor and removed unnecessary code (URL parsing classes). Also created CurlLoggingInterceptor, LogUtils class and an additional method in CoreUtils.

Maybe we can have a chat tomorrow to go over these changes @bsiegel. What do you think?

**Comment by anuchandy** [comment]
@bsiegel @vcolin7 slf4j logger used by the ClientLogger (our logger abstraction) redirect log calls to android logging system via Log.v and other variants. http://www.slf4j.org/android/

**Pull request merged**